spec-abstract-noclosure/ffg.tla

```diff
@@ -0,0 +1,265 @@
+         /\ chain2' = chain2 \union { new_block }
+         /\ chain2_tip' = new_block
+    /\ UNCHANGED <<ffg_votes, votes, justified_checkpoints, chain2_fork_block_number>>
+
+\* Append a block to chain 2.
+\* @type: Int => Bool;
+ProposeBlockOnChain2(slot) ==
+    LET new_block == Block(slot, NextBody(chain2_tip, "fork")) IN
+    /\ IsValidBlock(new_block)
+    /\ slot > chain2_tip.slot
+    /\ chain2_fork_block_number' =
+        IF IsForked THEN chain2_fork_block_number ELSE new_block.body
+    /\ all_blocks' = all_blocks \union { new_block }
+    /\ chain2' = chain2 \union { new_block }
+    /\ chain2_tip' = new_block
+    /\ UNCHANGED <<chain1, chain1_tip, ffg_votes, votes, justified_checkpoints>>
+
+JustifiedCheckpoints(viewVotes) ==
+    \* @type: Set($checkpoint) => Set($checkpoint);
+    LET AccJustified(justifiedSoFar, justifiedCheckpointSlot) ==
+        LET candidateCheckpoints == { Checkpoint(block, justifiedCheckpointSlot): block \in all_blocks } IN
+        LET newJustifiedCheckpoints == { c \in candidateCheckpoints: IsJustified(c, viewVotes, justifiedSoFar) } IN
+        justifiedSoFar \union newJustifiedCheckpoints
+    IN ApaFoldSeqLeft(AccJustified, { GenesisCheckpoint }, MkSeq(MAX_BLOCK_SLOT+2, (* @type: Int => Int; *) LAMBDA i: i))
+
+\* @type: ($checkpoint, $checkpoint, Set(Str)) => Bool;
+CastVotes(source, target, validators) ==
+    LET ffgVote == [ source |-> source, target |-> target ] IN
+    /\ IsValidFFGVote(ffgVote)
+    /\ validators /= {}
+    /\ ffg_votes' = ffg_votes \union { ffgVote }
+    /\ votes' = votes \union { Vote(v, ffgVote): v \in validators }
+    /\ LET allCheckpoints == {Checkpoint(block, i): block \in all_blocks, i \in CheckpointSlots}
+       IN \E allJustifiedCheckpoints \in SUBSET allCheckpoints:
+        /\ justified_checkpoints' = allJustifiedCheckpoints
+        /\ \A c \in allJustifiedCheckpoints: IsJustified(c, votes', allJustifiedCheckpoints)
+        /\ \A c \in (allCheckpoints \ allJustifiedCheckpoints): ~IsJustified(c, votes', allJustifiedCheckpoints)
+    \*/\ justified_checkpoints' = JustifiedCheckpoints(votes')
+    /\ UNCHANGED <<all_blocks, chain1, chain1_tip, chain2, chain2_tip, chain2_fork_block_number>>
+
+ExistTwoConflictingBlocks == \A b1, b2 \in all_blocks: ~AreConflictingBlocks(b1, b2)
+ExistTwoFinalizedConflictingBlocks ==
+    LET disagreement == \E c1, c2 \in justified_checkpoints: 
+        /\ IsFinalized(c1, votes, justified_checkpoints)
+        /\ IsFinalized(c2, votes, justified_checkpoints)
+        /\ AreConflictingBlocks(c1[1], c2[1])
+    IN ~disagreement
+
+AccountableSafety ==
+    LET disagreement == \E c1, c2 \in justified_checkpoints: 
+        /\ IsFinalized(c1, votes, justified_checkpoints)
+        /\ IsFinalized(c2, votes, justified_checkpoints)
+        /\ AreConflictingBlocks(c1[1], c2[1])
+    IN ~disagreement \/ Cardinality(SlashableNodes) * 3 >= N
+
+Init == 
+    /\ all_blocks = { GenesisBlock }
+    /\ chain1_tip = [ slot |-> 0, body |-> 0 ]
```

Comment:
I think we can write 

`/\ chain1_tip = GenesisBlock
  /\ chain2_tip = GenesisBlock`